File: Bifluidefinal/time-laplacian.cpp
/* -------------------------------------------------------------------------*\
 *
 *  NEOS
 *
 *  -------------------------------------------------------------------------
 *  License
 *  This file is part of Neos.
 *
 *  Neos is free software: you can redistribute it and/or modify it
 *  under the terms of the GNU Lesser General Public License v3 (LGPL)
 *  as published by the Free Software Foundation.
 *
 *  Neos is distributed in the hope that it will be useful, but WITHOUT
 *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 *  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 *  License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with Neos. If not, see <http://www.gnu.org/licenses/>.
 *
\*---------------------------------------------------------------------------*/

#include "Neos.hpp"
#include "LaplacianFactory.hpp"
#include <math.h>
#include <iostream>
#include <string>

#define TYPEGEOM 1
#define MAXRAF 1

using namespace neos;

int main(int ac, char **av) {

  Neos_Init(&ac,&av);

  int level = 4; // 4
  int dimension= 3; // 3
  int maxr= std::stoi(av[MAXRAF]);
  double temperature_core = 0.8;
  double temperature_border = 0.2;
  double temperature_ambiant = 0.2;
  double coef_penalization = 1.0e20; // 1.0e20
  double tolerance_penalization = 0.1; // 0.1
  double coef_refinment = 0.2;
  double delta_t= 0.001; // 0.03; // time step
  int time_iteration= 90;
  int time_enable= 1; // 0: disable, 1: enable time step
  int border_enable= 0; // 0 disable, 1: enable border penalisation

  double h=1;  // grid step -

  Grid *g;
  LevelSet *ls;

  if (!time_enable)
  {
    time_iteration= 1;
  }

  if (ac != 2) {
    std::cerr << av[0] << " 1 for ASphere, 2 for STL" << std::endl;
    exit(1);
  }

  int typeGeom = std::stoi(av[TYPEGEOM]);

  int geoId = -1;

  switch (typeGeom) {
    case 1: {
      g = new Grid(0.0, 0.0, 0.0, 1.0, 1.0/pow(2,level), dimension);
      g->update(false,true);
      ASphere *geoSphere  = new ASphere(0.5, 0.5, 0.5, 0.2, dimension);
      geoSphere->computeLevelSet(g);
      ls = new LevelSet(g);
      geoId = ls->addGeometry(geoSphere, "TAG");
      break;
    }
    default: {
      g = new Grid(-20, -20, -10, 50.0, 2.0, dimension);
      STLGeometry *geoSTL = new STLGeometry();
      geoSTL->addSTL("./Dragonite.stl", g);
      ls = new LevelSet(g);
      geoSTL->computeLevelSet(g);
      geoId = ls->addGeometry(geoSTL, "TAG");
      delta_t= 0.1;
      coef_refinment = 20.0;
    }
  }

  std::vector<double> phi0 = ls->getLevelSet("TAG");

  int nbRef = g->refineFromRangeVector(phi0,0,coef_refinment);
  std::cout << nbRef << " Cells Refined" << '\n';
  Geometry *geom = ls->getGeometry(geoId);
  geom->computeLevelSet(g);
  phi0 = ls->getLevelSet("TAG");

  double h_h= pow(h,dimension);
  double h_h_sur_delta_t= h_h/delta_t;

  std::cerr << " h : " << h << " h h " << h_h << " delta_t " << delta_t << " h_h_sur_delta_t " << h_h_sur_delta_t << std::endl;

  std::vector<double> pen = phi0;
  std::vector<double> initial_temperature_RHS = phi0;

  for (auto &element : pen) {
    if (element > tolerance_penalization) {
      element = 0.0; // no penalization
    } else {
      element = 1.0; // penalization inside
    }
  }
  std::vector<double> fixed_temperature_RHS = pen * temperature_core;

  for (auto &element : initial_temperature_RHS) {
    if (element > tolerance_penalization) {
      element = temperature_ambiant;
    } else {
      element = temperature_core;
    }
  }

  std::vector<double> sol_v= initial_temperature_RHS;

  if (0) {
    for (auto &element : sol_v) {
      element = 0.0;
    }
  }

  std::vector<std::array<double, 3> > gradvtk;

  g->setExportName("time-laplacian");
  g->setCounter();

  Laplacian *lap = (Laplacian *)LaplacianFactory::get(lapType::FINITEVOLUME, solverType::PETSC, g);

  lap->setBCPositionCoef(BCPosition::OnInterface);
  lap->buildMatrix();

  lap->penalize(pen * coef_penalization * (-1.0) * h_h - time_enable * h_h_sur_delta_t);

  for (int i=0; i<time_iteration; i++) {
    lap->setRHS( fixed_temperature_RHS * coef_penalization * (-1.0) * h_h - time_enable * h_h_sur_delta_t * sol_v); // RH

    double norm = lap->solveLaplacian();
    sol_v = lap->getSolution();
    std::cout << "Iter " << i << " Norm " << norm << std::endl;

    g->addData("solution",sol_v);
    g->addData("phi",phi0);

    g->write();
  }
  delete g;

  Neos_Finalize();
}

File: Bifluidefinal/time-laplacian.py
# -------------------------------------------------------------------------
#
#  NEOS
#
#  -------------------------------------------------------------------------
#  License
#  This file is part of Neos.
#
#  Neos is free software: you can redistribute it and/or modify it
#  under the terms of the GNU Lesser General Public License v3 (LGPL)
#  as published by the Free Software Foundation.
#
#  Neos is distributed in the hope that it will be useful, but WITHOUT
#  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
#  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
#  License for more details.
#
#  You should have received a copy of the GNU Lesser General Public License
#  along with Neos. If not, see <http://www.gnu.org/licenses/>.
#
#---------------------------------------------------------------------------

from pyneos.pyneos import neos
from mpi4py import MPI
import sys
import numpy as np
from pprint import pprint

#help(neos)
print(dir(neos.Grid))

level = 4
dimension= 3
temperature_core = 0.8
temperature_border = 0.2
temperature_ambiant = 0.2
coef_penalization = 1.0e20
tolerance_penalization = 0.1
coef_refinment = 0.2
delta_t= 0.001
time_iteration= 90
time_enable= 1
border_enable= 0
h=1

if (sys.argv[1] == "1"):
    grid = neos.Grid(0, 0, 0, 1, 1.0/pow(2,level), dimension)
    geo = neos.ASphere(0.5, 0.5, 0.5, 0.2, dimension)
    geo.computeLevelSet(grid);
    ls = neos.LevelSet(grid);
    geoId = ls.addGeometry(geo)
else:
    grid = neos.Grid(-20, -20, -10, 50.0, 2, dimension)
    geo = neos.STLGeometry()
    geo.addSTL("./Dragonite.stl", grid)
    geo.computeLevelSet(grid)
    ls = neos.LevelSet(grid)
    geoId = ls.addGeometry(geo)
    delta_t= 0.1
    coef_refinment = 10.0

phi0 = ls.getLevelSet()

nbRef = grid.refineFromRangeVector(phi0,0.0,coef_refinment);

print(nbRef," Cells Refined")

geo.computeLevelSet(grid)
phi0 = ls.getLevelSet()
pen = np.array(phi0)
initial_temperature_RHS = np.array(phi0)

h_h= pow(h,dimension)
h_h_sur_delta_t= h_h/delta_t

for i in range(len(pen)):
  if (pen[i] > tolerance_penalization):
    pen[i] = 0.0
  else:
    pen[i] = 1.0

fixed_temperature_RHS = pen * temperature_core

for i in range(len(initial_temperature_RHS)):
    if initial_temperature_RHS[i] > tolerance_penalization:
        initial_temperature_RHS[i] = temperature_ambiant
    else:
        initial_temperature_RHS[i] = temperature_core

sol_v = np.array(initial_temperature_RHS)

grid.setExportName("time-laplacian")
grid.setCounter()

lap = neos.LaplacianFactory.get(neos.lapType.FINITEVOLUME, neos.solverType.PETSC, grid);

#lap.setBCPositionCoef(neos.BCPosition.OnInterface);
lap.buildMatrix();
lap.penalize(pen * coef_penalization * (-1.0) * h_h - time_enable * h_h_sur_delta_t);

for i in range(time_iteration):
    lap.setRHS( fixed_temperature_RHS * (coef_penalization * (-1.0) * h_h) -
               (time_enable * h_h_sur_delta_t) * sol_v)
    norm = lap.solveLaplacian()
    print("Iter ", i,  " Norm ", norm)
    sol_v = lap.getSolution()
    lap.addSolutionToVTK("solution");
    grid.addData("phi",phi0)
    grid.write()
    sol_v = np.array(sol_v)
